src/widgets/souvenirlist.cpp

```diff
@@ -0,0 +1,294 @@
+    }
+}
+
+/**
+ * @brief Remove souvenir from the list
+ *
+ * Searches through the list and looks for matching IDs by casting
+ * each widget into a @a SouvenirListItem. When found, the souvenir
+ * is removed from the IDs-quantity map as well as the list.
+ *
+ * @param id IDs of the souvenir to remove
+ */
+void SouvenirList::removeItem(IDs id)
+{
+    for(int row = 0; row < QListWidget::count(); row++)
+    {
+        SouvenirListItem* widget = castRow(row);
+
+        if(widget != nullptr && id == widget->getIDs())
+        {
+            m_IDQtys.erase(widget->getIDs());
+            QListWidget::takeItem(row);
+            return;
+        }
+    }
+}
+
+/**
+ * Each item's quantity spinbox will either hide or show depending
+ * on @a show. This is useful if you need to input quantity for each item.
+ *
+ * @param show Show each item's quantity spinbox
+ */
+void SouvenirList::showQty(bool show)
+{
+    m_showQty = show;
+
+    for(int row = 0; row < QListWidget::count(); row++)
+    {
+        SouvenirListItem* widget = castRow(row);
+
+        if(widget != nullptr) { widget->showQty(show); }
+    }
+}
+
+/**
+ * Clears out the container holding the IDs-quantity relationship
+ * and changes resets each widget's quantity.
+ */
+void SouvenirList::resetQty()
+{
+    m_IDQtys.clear();
+
+    for(int row = 0; row < QListWidget::count(); row++)
+    {
+        SouvenirListItem* widget = castRow(row);
+
+        if(widget != nullptr) { widget->resetQty(); }
+    }
+}
+
+/**
+ * Sets whether or not hidden souvenirs will be added or not.
+ *
+ * NOTE:
+ * Calling this function after items are added will have
+ * no affect on those items.
+ *
+ * @param allow Allow hidden souvenirs to show
+ */
+void SouvenirList::allowHidden(bool allow)
+{
+    m_allowHidden = allow;
+}
+
+/**
+ * Converts a row in the list to IDs and emits a signal with those IDs.
+ *
+ * @param row Row of the item to convert into IDs
+ */
+void SouvenirList::rowToIDsEmitter(int row) const
+{
+    SouvenirListItem* widget = castRow(row);
+
+    if(widget != nullptr)
+    {
+        emit currentSouvenirChanged(widget->getIDs());
+    }
+}
+
+/**
+ * When a quantity spinbox has changed value, this function will
+ * store that new value. If the changing quantity is the minimum
+ * quantity, the changed item is removed from the list.
+ *
+ * @param id IDs of the changed souvenir
+ * @param qty New quantity value
+ */
+void SouvenirList::quantityChangedHandler(IDs id, int qty)
+{
+    if(qty != m_minQty)
+    {
+        m_IDQtys[id] = qty; //Store/replace the key with the value
+    }
+    else
+    {
+        removeItem(id); //Remove it from the list
+    }
+}
+
+/**
+ * Since @a QListWidgetItem can hold a pointer to a @a QWidget, we
+ * can dynamic cast the @a QWidget that the @a QListWidgetItem
+ * holds into a @a SouvenirListItem.
+ *
+ * That is exactly what this function does: dynamic casts
+ * a @a QListWidgetItem at a given row and the return the result.
```

Comment:
Say that it **dynamic casts a QListWidgetItem at a given row into a SouvenirListItem and then returns the result**